src/Data/Text/Utf8.hs

```diff
@@ -46,49 +43,80 @@ module Data.Text.Utf8
 import Data.Bits (Bits (shiftL), shiftR, (.&.), (.|.))
 import Data.Char (ord)
 import Data.Foldable (for_)
 import Data.Hashable (Hashable (hashWithSalt), hashByteArrayWithSalt)
-import Data.Primitive.ByteArray (ByteArray (ByteArray), byteArrayFromList, indexByteArray,
-                                 newByteArray, sizeofByteArray, unsafeFreezeByteArray,
-                                 writeByteArray)
+import Data.Primitive.ByteArray (ByteArray (ByteArray), byteArrayFromList, compareByteArrays,
+                                 indexByteArray, newByteArray, sizeofByteArray,
+                                 unsafeFreezeByteArray, writeByteArray)
+import Data.String (IsString (fromString))
 import Data.Word (Word8)
-import GHC.Base (Int (I#), compareByteArrays#)
 import GHC.Generics (Generic)
 import Prelude hiding (length)
 #if defined(HAS_AESON)
 import Data.Aeson (FromJSON, ToJSON, Value (String), parseJSON, toJSON, withText)
 #endif
 
-import qualified Data.Aeson as AE
 import qualified Data.ByteString as BS
 import qualified Data.Char as Char
 import qualified Data.Text as T
 
+-- | A UTF-8 code unit is a byte. A Unicode code point can be encoded as up to four code units.
 type CodeUnit = Word8
+
+-- | A Unicode code point.
 type CodePoint = Char
 
+-- | An index into the raw UTF-8 data of a `Text`. This is not the code point
+-- index as conventionally accepted by `Text`, so we wrap it to avoid confusing
+-- the two. Incorrect index manipulation can lead to surrogate pairs being
+-- sliced, so manipulate indices with care. This type is also used for lengths.
+newtype CodeUnitIndex = CodeUnitIndex
+    { codeUnitIndex :: Int
+    }
+    deriving stock (Eq, Ord, Show, Generic, Bounded)
+#if defined(HAS_AESON)
+    deriving newtype (Hashable, Num, NFData, FromJSON, ToJSON)
+#else
+    deriving newtype (Hashable, Num, NFData)
+#endif
+
 data Text
   -- | A placeholder data type for UTF-8 encoded text until we can use text-2.0.
   = Text
       !ByteArray -- ^ Underlying array encoded using UTF-8.
       !Int -- ^ Starting position of the UTF-8 sequence in bytes.
       !Int -- ^ Length of the UTF-8 sequence in bytes.
-  deriving Show
 
 -- This instance, as well as the Show instance above, is necessary for the test suite.
 instance Eq Text where
-  -- Since Data.Primitive.ByteArray.compareByteArrays is not available in lts-16.28, use the primitive version
-  -- from GHC.Base
-  (Text (ByteArray u8data) (I# offset) (I# length)) == (Text (ByteArray u8data') (I# offset') (I# length')) =
-    I# length == I# length' && I# (compareByteArrays# u8data offset u8data' offset' length) == 0
+  (Text u8data offset length) == (Text u8data' offset' length') =
+    length == length' && compareByteArrays u8data offset u8data' offset' length == EQ
+
+instance Ord Text where
+  compare (Text u8data offset length) (Text u8data' offset' length') =
+    compare length length' <> compareByteArrays u8data offset u8data' offset' (min length length')
```

Comment:
While this probably results in a valid Ord instance, it doesn't have an ordering that one expects form ordering texts I think: since it checks length first, `"X" < "ABC"`, but lexicographic ordering (which is more sensible) is `"ABC" < "X"`.

Reply:
Addressed in 3c83d97. I also added tests for `Show Text`, `Eq Text` and `Ord Text` to make sure that they behave the same as the corresponding `String` instances.